feat(rust): `ockam node show` to use dynamic data from node

neil2468 opened the pull request:
## Current Behavior

* `ockam node show` intends to show the status and details of a node (e.g. a list of its services). At the moment this feature uses static text and assumes things about the node (see below code).
* This code is also used by `ockam node list` and `ockam node create`.
* Related to #3177.

https://github.com/build-trust/ockam/blob/d4dd3529e09139dbebebbf8ba59db114353e8330/implementations/rust/ockam/ockam_command/src/node/show.rs#L43-L79

## Proposed Changes

* Modify `ockam node show` to ask the node for details about itself and display those details.
* Add a request/response `get` `/node/service` to allow the CLI to get a list of services from a node.
* Add a basic test to `ockam_command/tests/commands/bats` to test the new feature in CI.

## Queries

1. Is the output formatting of `ockam node show` okay? Instead of a verbose tree, could we output some of the information in a tabular style like `docker image ls`?

```shell
> docker image ls
REPOSITORY                                     TAG       IMAGE ID       CREATED        SIZE
ghcr.io/build-trust/ockam-builder              latest    b10ae04c8b6f   2 weeks ago    4.62GB
ghcr.io/build-trust/ockam-builder              <none>    9cfba9ba6e22   2 months ago   4.58GB
ghcr.io/cross-rs/riscv64gc-unknown-linux-gnu   0.2.4     23b40ea6412a   3 months ago   1.03GB
hello-world                                    latest    46331d942d63   6 months ago   9.14kB
```

2. In the output, should the Transport field `Payload` be renamed?

3. The request/response `get` `/node/service` builds its response from `ockam_api::nodes::service::NodeManager.registry`. Is there a risk that this might leak private information from the registry to a user using the CLI?

4. Are there other details about a node that should be added to this feature (e.g. TCP inlets and outlets)?

## Example Output

```shell
> ockam node create n1
> ockam node create n2
> ockam secure-channel-listener create test --at /node/n1
> ockam tcp-listener create 127.0.0.1:7002 --at /node/n1
> ockam node list

Node:
  Name: n1
  Status: UP
  Route to node:
    Short: /node/n1
    Verbose: /dnsaddr/localhost/tcp/55754
  Identity: Pcf94e075af7c5885ab0eb84abb9a76223f7359f6c64bf753a136ae1b2bb3f02f
  Transports:
    Transport:
      Type: TCP
      Mode: Listening
      Payload: 127.0.0.1:55754
    Transport:
      Type: TCP
      Mode: Listening
      Payload: 127.0.0.1:7002
  Secure Channel Listeners:
    Listener:
      Address: /service/api
    Listener:
      Address: /service/test
  Services:
    Service:
      Type: vault
      Address: /service/vault_service
    Service:
      Type: identity
      Address: /service/identity_service
    Service:
      Type: authenticated
      Address: /service/authenticated
    Service:
      Type: uppercase
      Address: /service/uppercase
    Service:
      Type: echoer
      Address: /service/echo
    Service:
      Type: credentials
      Address: /service/credentials

Node:
  Name: n2
  Status: UP
  Route to node:
    Short: /node/n2
    Verbose: /dnsaddr/localhost/tcp/55761
  Identity: Pcf94e075af7c5885ab0eb84abb9a76223f7359f6c64bf753a136ae1b2bb3f02f
  Transports:
    Transport:
      Type: TCP
      Mode: Listening
      Payload: 127.0.0.1:55761
  Secure Channel Listeners:
    Listener:
      Address: /service/api
  Services:
    Service:
      Type: vault
      Address: /service/vault_service
    Service:
      Type: identity
      Address: /service/identity_service
    Service:
      Type: authenticated
      Address: /service/authenticated
    Service:
      Type: uppercase
      Address: /service/uppercase
    Service:
      Type: echoer
      Address: /service/echo
    Service:
      Type: credentials
      Address: /service/credentials
```

## Checks

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository.

neil2468 committed:
feat(rust): `ockam node show` to use dynamic data from node

mrinalwadhwa reviewed (dismissed):
That looks great! 
Thank you for another awesome contribution!

mrinalwadhwa commented:
> Is the output formatting of ockam node show okay? Instead of a verbose tree, could we output some of the information in a tabular style like docker image ls?

This is a good question to consider, let's do that in a separate issue. The main thing to discuss would be what information makes sense in the table like output.

> In the output, should the Transport field Payload be renamed?

Yes it should be called Address.

> The request/response get /node/service builds its response from ockam_api::nodes::service::NodeManager.registry. Is there a risk that this might leak private information from the registry to a user using the CLI?

No, I think.

> Are there other details about a node that should be added to this feature (e.g. TCP inlets and outlets)?

Yes. Inlets & Outlets should be there. I'll think about what else.

neil2468 commented:
Hi,

I'm looking at making some improvements and adding the inlets and outlets. 
Sorry for not making the PR draft earlier 🙏 

Thanks, Neil.

mrinalwadhwa commented:
Great, glad you noticed I was about to merge 😬

neil2468 commented:
# Additional Proposed Changes

* Inside a `NodeManager`'s `registry`, `InletInfo` stores the onward route to the inlet's outlet.
* `ockam node show` lists the details of a node's inlets and outlets.

# Queries

1. I removed the comment `// FIXME route.as_ref()...` from `ockam_api/src/nodes/service/portals.rs`. Is this okay? @SanjoDeundiak 

2. I added the function `addr_to_multiaddr<T: Into<Address>>(a: T) -> Option<MultiAddr>`. As it is public, is its API okay?

# New Example Output

```shell
> ockam node create n1
> ockam node create n2
> ockam tcp-outlet create --at /node/n1 --from /service/outlet --to 127.0.0.1:5000
> ockam tcp-inlet create --at /node/n2 --from 127.0.0.1:6000 --to /node/n1/service/outlet
> ockam node list


Node:
  Name: n1
  Status: UP
  Route To Node:
    Short: /node/n1
    Verbose: /dnsaddr/localhost/tcp/58434
  Identity: P5f2178c1852c02f602ead6f261133f036b1e4910058cc9c6cdfcddd5605d87cd
  Transports:
    Transport:
      Type: TCP
      Mode: Listening
      Address: 127.0.0.1:58434
  Secure Channel Listeners:
    Listener:
      Address: /service/api
  Inlets:
  Outlets:
    Outlet:
      Forward Address: 127.0.0.1:5000
      Address: /service/outlet
  Services:
    Service:
      Type: vault
      Address: /service/vault_service
    Service:
      Type: identity
      Address: /service/identity_service
    Service:
      Type: authenticated
      Address: /service/authenticated
    Service:
      Type: uppercase
      Address: /service/uppercase
    Service:
      Type: echoer
      Address: /service/echo
    Service:
      Type: credentials
      Address: /service/credentials

Node:
  Name: n2
  Status: UP
  Route To Node:
    Short: /node/n2
    Verbose: /dnsaddr/localhost/tcp/58441
  Identity: P5f2178c1852c02f602ead6f261133f036b1e4910058cc9c6cdfcddd5605d87cd
  Transports:
    Transport:
      Type: TCP
      Mode: Listening
      Address: 127.0.0.1:58441
  Secure Channel Listeners:
    Listener:
      Address: /service/api
  Inlets:
    Inlet:
      Listen Address: 127.0.0.1:6000
      Route To Outlet: /ip4/127.0.0.1/tcp/58434/service/outlet
  Outlets:
  Services:
    Service:
      Type: vault
      Address: /service/vault_service
    Service:
      Type: identity
      Address: /service/identity_service
    Service:
      Type: authenticated
      Address: /service/authenticated
    Service:
      Type: uppercase
      Address: /service/uppercase
    Service:
      Type: echoer
      Address: /service/echo
    Service:
      Type: credentials
      Address: /service/credentials
```

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/show.rs:
```diff
@@ -86,59 +149,138 @@ pub async fn print_query_status(
-        .context("Failed to process request");
+    for att in 0..attempts {
+        // Sleep, if this not the first loop
+        if att > 0 {
+            tokio::time::sleep(Duration::from_millis(IS_NODE_UP_SLEEP_MILLIS)).await;
+        }
```
Wouldn't we have the same behavior if the `sleep` is called at the end of the loop? That way we could remove the `if` block.

neil2468 replied:
I want to avoid sleeping if the caller only wants us to do one attempt.

If the `sleep` were moved to the end of the loop, I think it would still need an `if`. Consider when `wait_until_ready=false`and the node is down, without an `if` there would be a sleep before returning.

Thanks!

adrianbenavides replied:
Yeah, true, I didn't consider that case. Thanks for the clarification!

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/show.rs:
```diff
@@ -86,59 +149,138 @@ pub async fn print_query_status(
+            .await;
+        if let Ok(data) = tx_result {
+            if api::parse_status(&data).is_ok() {
+                // Node is up, break loop
+                node_up = true;
+                break;
```
Given that the only exit condition for this loop is when the node is up, maybe you could `return Ok(true)` here and `Ok(false)` at the end of the function, so we can get rid of the variable that tracks the node status.

neil2468 replied:
Nice. Simplifying code make me happy. Thanks!

adrianbenavides reviewed (commented):
That node output looks pretty great! I left a couple of nits, @SanjoDeundiak feel free to merge it whenever you want!

adrianbenavides commented:
@mergifyio rebase

adrianbenavides commented:
@neil2468 looks like your commits are not signed. Could you please force-push your last commit with your signature? Here you can find more details on how to do it: https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

neil2468 commented:
@adrianbenavides Done! Thanks.
The Rust Nightly CI fails on my fork and I've noticed the same is happening in the Ockam repo. 
Rust stable seems okay 😄.

adrianbenavides commented:
> The Rust Nightly CI fails on my fork and I've noticed the same is happening in the Ockam repo. Rust stable seems okay smile.

Yeah, don't worry about that. We looked into that and it seems a [bug on the latest nightly](https://github.com/rust-lang/rust/pull/102567) version.

adrianbenavides reviewed (dismissed):
LGTM!! :rocket: Thanks again @neil2468, this PR is a great improvement :pray:

adrianbenavides commented:
Here too @neil2468, although this time looks like it's a simple conflict. I'll merge it as soon as you push your changes!

mrinalwadhwa reviewed (approved):
Thank you for another awesome contribution! 🙏
